Os/Linux/SystemResources.cpp

```diff
@@ -9,107 +9,159 @@
+        return cpu_data[0] + cpu_data[1] + cpu_data[2];
+    }
+    U64 getCpuTotal(std::array<U32, 4>& cpu_data) {
+        return cpu_data[0] + cpu_data[1] + cpu_data[2] + cpu_data[3];
+    }
 
-        if ((status = getCpuCount(cpuCount)) != SYSTEM_RESOURCES_OK) {
-            return status;
+    SystemResources::SystemResourcesStatus openProcStatFile(FILE*& fp) {
+        if ((fp = fopen(PROC_STAT_PATH, READ_ONLY)) == nullptr) {
+            return SystemResources::SYSTEM_RESOURCES_ERROR;
         }
+        return SystemResources::SYSTEM_RESOURCES_OK;
+    }
 
-        if (cpu_index >= cpuCount) {
-            return SYSTEM_RESOURCES_ERROR;
+    SystemResources::SystemResourcesStatus readProcStatLine(FILE* fp, std::array<char, LINE_SIZE>& proc_stat_line) {
+        if (fgets(proc_stat_line.data(), proc_stat_line.size(), fp) == nullptr) {
+            return SystemResources::SYSTEM_RESOURCES_ERROR;
         }
+        return SystemResources::SYSTEM_RESOURCES_OK;
+    }
 
-        if ((fp = fopen("/proc/stat", "r")) == nullptr) {
-
-            return SYSTEM_RESOURCES_ERROR;
+    SystemResources::SystemResourcesStatus getCpuDataLine(FILE* fp,
+                                                          U32 cpu_index,
+                                                          std::array<char, LINE_SIZE>& proc_stat_line) {
+        for (U32 i = 0; i < cpu_index + 1; i++) {
+            if (readProcStatLine(fp, proc_stat_line) != SystemResources::SYSTEM_RESOURCES_OK) {
+                return SystemResources::SYSTEM_RESOURCES_ERROR;
+            }
+            if (i != cpu_index) {
+                continue;
+            }
+            if (strncmp(proc_stat_line.data(), "cpu", 3) != 0) {
+                return SystemResources::SYSTEM_RESOURCES_ERROR;
+            }
+            break;
+        }
+        return SystemResources::SYSTEM_RESOURCES_OK;
+    }
 
+    SystemResources::SystemResourcesStatus parseCpuData(std::array<char, LINE_SIZE>& proc_stat_line,
+                                                        std::array<U32, 4>& cpu_data) {
+        if (sscanf(proc_stat_line.data(), "%*s %d %d %d %d", &cpu_data[0], &cpu_data[1], &cpu_data[2], &cpu_data[3]) !=
+            4) {
+            return SystemResources::SYSTEM_RESOURCES_ERROR;
         }
+        return SystemResources::SYSTEM_RESOURCES_OK;
+    }
 
-        if (fgets(line, sizeof(line), fp) == nullptr) { //1st line.  Aggregate cpu line.
+    SystemResources::SystemResourcesStatus getCpuData(U32 cpu_index, std::array<U32, 4>& cpu_data) {
+        FILE* fp = nullptr;
+        if (openProcStatFile(fp) != SystemResources::SYSTEM_RESOURCES_OK ||
+            readProcStatLine(fp, proc_stat_line) != SystemResources::SYSTEM_RESOURCES_OK ||
+            getCpuDataLine(fp, cpu_index, proc_stat_line) != SystemResources::SYSTEM_RESOURCES_OK ||
+            parseCpuData(proc_stat_line, cpu_data) != SystemResources::SYSTEM_RESOURCES_OK) {
             fclose(fp);
```

Comment:
Note: I have to create a `closeProcStatFile` function to have the same level of abstraction as its antagonist `openProcStatFile`.